diff --git a/node_view_permissions.module b/node_view_permissions.module
index 2a28f2e..5e07486 100644
--- a/node_view_permissions.module
+++ b/node_view_permissions.module
@@ -112,24 +112,25 @@ function node_view_permissions_query_node_access_alter(QueryAlterableInterface $
     return;
   }
 
+  $node_table_alias = FALSE;
   //if viewing the node, deny access if the type is a configured type and no node grants allow access
   if ($op == 'view') {
     $tables = $query->getTables();
     // @see _node_query_node_access_alter().
-    $node_table_alias = FALSE;
     foreach ($tables as $alias => $table) {
       if ($table['table'] == 'node') {
         $node_table_alias = $alias;
         break;
       }
     }
-    
+
     if (!$node_table_alias) {
+      reset($tables);
       // If the query has 'node_access' meta tag, we can assume that there should be a nid field.
       $query->join('node', 'n', 'n.nid = ' . key($tables) . '.nid');
       $node_table_alias = 'n';
     }
-    
+
     if ($node_table_alias) {
       $alias = $node_table_alias;
       foreach (node_view_permissions_get_configured_types() as $type) {
